File: examples/client-example.cs
/*
 * FZ Auth — C# Client Example
 *
 * Works with .NET 6+ / .NET Framework 4.7.2+
 * Add this class to your project and call FZAuth.Initialize(), FZAuth.Login(), etc.
 *
 * Usage:
 *   1. Create an app in the FZ Auth panel and copy app_id + secret
 *   2. Generate a license key in the panel
 *   3. Set the constants below
 */

using System;
using System.Collections.Generic;
using System.Management;
using System.Net.Http;
using System.Text;
using System.Text.Json;
using System.Threading.Tasks;

public static class FZAuth
{
    // ── Replace these with your real values from the panel ──────────
    private const string BASE       = "http://localhost:3001";
    private const string APP_ID     = "YOUR_APP_ID_HERE";
    private const string APP_SECRET = "YOUR_APP_SECRET_HERE";
    private const string VERSION    = "1.0";
    // ────────────────────────────────────────────────────────────────

    private static readonly HttpClient _http = new HttpClient();

    public static string Token { get; private set; }
    public static string Username { get; private set; }
    public static string Expiry { get; private set; }
    public static string HWID => GetHWID();

    private static async Task<JsonElement> Api(string endpoint, Dictionary<string, object> extra = null)
    {
        var body = new Dictionary<string, object>
        {
            ["app_id"] = APP_ID,
            ["secret"] = APP_SECRET
        };

        if (extra != null)
            foreach (var kv in extra)
                body[kv.Key] = kv.Value;

        var json = JsonSerializer.Serialize(body);
        var content = new StringContent(json, Encoding.UTF8, "application/json");
        var res = await _http.PostAsync($"{BASE}/api/v1/{endpoint}", content);
        var raw = await res.Content.ReadAsStringAsync();
        return JsonSerializer.Deserialize<JsonElement>(raw);
    }

    private static string GetHWID()
    {
        try
        {
            using var searcher = new ManagementObjectSearcher("SELECT ProcessorId FROM Win32_Processor");
            foreach (var obj in searcher.Get())
                return obj["ProcessorId"]?.ToString() ?? Environment.MachineName;
        }
        catch { }
        return Environment.MachineName + "-" + Environment.UserName;
    }

    public static async Task<bool> Initialize()
    {
        var data = await Api("init", new() { ["version"] = VERSION });
        if (!data.GetProperty("success").GetBoolean())
        {
            Console.WriteLine($"Init failed: {data.GetProperty("message")}");
            return false;
        }
        Console.WriteLine($"App: {data.GetProperty("app_name")} v{data.GetProperty("version")}");
        return true;
    }

    public static async Task<bool> Login(string username, string password)
    {
        var data = await Api("login", new()
        {
            ["username"] = username,
            ["password"] = password,
            ["hwid"] = HWID
        });

        if (!data.GetProperty("success").GetBoolean())
        {
            Console.WriteLine($"Login failed: {data.GetProperty("message")}");
            return false;
        }

        Token = data.GetProperty("token").GetString();
        Username = data.GetProperty("username").GetString();
        Expiry = data.GetProperty("expiry").GetString();
        return true;
    }

    public static async Task<bool> Register(string username, string password, string licenseKey)
    {
        var data = await Api("register", new()
        {
            ["username"] = username,
            ["password"] = password,
            ["license_key"] = licenseKey,
            ["hwid"] = HWID
        });

        if (!data.GetProperty("success").GetBoolean())
        {
            Console.WriteLine($"Register failed: {data.GetProperty("message")}");
            return false;
        }

        Token = data.GetProperty("token").GetString();
        Username = data.GetProperty("username").GetString();
        Expiry = data.GetProperty("expiry").GetString();
        return true;
    }

    public static async Task<bool> LicenseAuth(string licenseKey)
    {
        var data = await Api("license", new()
        {
            ["license_key"] = licenseKey,
            ["hwid"] = HWID
        });

        if (!data.GetProperty("success").GetBoolean())
        {
            Console.WriteLine($"License auth failed: {data.GetProperty("message")}");
            return false;
        }

        Token = data.GetProperty("token").GetString();
        Expiry = data.GetProperty("expiry").GetString();
        return true;
    }

    public static async Task<bool> CheckSession()
    {
        if (string.IsNullOrEmpty(Token)) return false;

        var data = await Api("check", new() { ["token"] = Token });
        return data.GetProperty("success").GetBoolean();
    }

    public static async Task<string> GetVariable(string key)
    {
        var data = await Api("var", new() { ["key"] = key });
        if (!data.GetProperty("success").GetBoolean()) return null;
        return data.GetProperty("variable").GetProperty("value").GetString();
    }

    public static async Task<Dictionary<string, string>> GetAllVariables()
    {
        var data = await Api("var");
        var result = new Dictionary<string, string>();
        if (!data.GetProperty("success").GetBoolean()) return result;
        foreach (var prop in data.GetProperty("variables").EnumerateObject())
            result[prop.Name] = prop.Value.GetString();
        return result;
    }
}

// ── Example usage ──────────────────────────────────────────────────
// class Program
// {
//     static async Task Main()
//     {
//         if (!await FZAuth.Initialize()) return;
//
//         // Login with username/password
//         if (!await FZAuth.Login("myuser", "mypass")) return;
//         Console.WriteLine($"Welcome {FZAuth.Username}! Expires: {FZAuth.Expiry}");
//
//         // OR register with license key
//         // if (!await FZAuth.Register("newuser", "pass123", "XXXX-XXXX-XXXX")) return;
//
//         // OR license-only auth
//         // if (!await FZAuth.LicenseAuth("XXXX-XXXX-XXXX")) return;
//
//         // Check session
//         bool valid = await FZAuth.CheckSession();
//         Console.WriteLine($"Session valid: {valid}");
//
//         // Fetch variables
//         var version = await FZAuth.GetVariable("update_url");
//         Console.WriteLine($"Update URL: {version}");
//     }
// }

File: examples/client-example.cpp
/*
 * FZ Auth — C++ Client Example (uses libcurl + nlohmann/json)
 *
 * Dependencies:
 *   - libcurl: https://curl.se/libcurl/
 *   - nlohmann/json: https://github.com/nlohmann/json (single header)
 *
 * Compile (Windows):
 *   g++ -o fzauth client-example.cpp -lcurl -lws2_32 -std=c++17
 *
 * Compile (Linux/macOS):
 *   g++ -o fzauth client-example.cpp -lcurl -std=c++17
 *
 * Usage:
 *   1. Create an app in the FZ Auth panel and copy app_id + secret
 *   2. Generate a license key in the panel
 *   3. Set the constants below and compile + run
 */

#include <iostream>
#include <string>
#include <curl/curl.h>
#include "json.hpp" // nlohmann/json single header

#ifdef _WIN32
  #include <winsock2.h>
  #pragma comment(lib, "ws2_32.lib")
#else
  #include <unistd.h>
#endif

using json = nlohmann::json;

// ── Replace these with your real values from the panel ──────────────
const std::string BASE       = "http://localhost:3001";
const std::string APP_ID     = "YOUR_APP_ID_HERE";
const std::string APP_SECRET = "YOUR_APP_SECRET_HERE";
const std::string VERSION    = "1.0";
// ────────────────────────────────────────────────────────────────────

static size_t WriteCallback(void* contents, size_t size, size_t nmemb, std::string* out) {
    out->append(static_cast<char*>(contents), size * nmemb);
    return size * nmemb;
}

json api(const std::string& endpoint, json body = {}) {
    body["app_id"] = APP_ID;
    body["secret"] = APP_SECRET;

    CURL* curl = curl_easy_init();
    std::string response;

    if (!curl) {
        return json{{"success", false}, {"message", "Failed to init curl"}};
    }

    std::string url = BASE + "/api/v1/" + endpoint;
    std::string payload = body.dump();

    struct curl_slist* headers = nullptr;
    headers = curl_slist_append(headers, "Content-Type: application/json");

    curl_easy_setopt(curl, CURLOPT_URL, url.c_str());
    curl_easy_setopt(curl, CURLOPT_POSTFIELDS, payload.c_str());
    curl_easy_setopt(curl, CURLOPT_HTTPHEADER, headers);
    curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, WriteCallback);
    curl_easy_setopt(curl, CURLOPT_WRITEDATA, &response);
    curl_easy_setopt(curl, CURLOPT_TIMEOUT, 10L);

    CURLcode res = curl_easy_perform(curl);
    curl_slist_free_all(headers);
    curl_easy_cleanup(curl);

    if (res != CURLE_OK) {
        return json{{"success", false}, {"message", std::string("Request failed: ") + curl_easy_strerror(res)}};
    }

    try {
        return json::parse(response);
    } catch (...) {
        return json{{"success", false}, {"message", "Invalid JSON response"}};
    }
}

std::string getHWID() {
    char name[256] = {0};
#ifdef _WIN32
    WSADATA wsaData;
    WSAStartup(MAKEWORD(2, 2), &wsaData);
    gethostname(name, sizeof(name));
    WSACleanup();
#else
    gethostname(name, sizeof(name));
#endif
    return std::string(name) + "-cpp-client";
}

int main() {
    curl_global_init(CURL_GLOBAL_ALL);
    std::cout << "FZ Auth - C++ Client\n\n";

    std::string hwid = getHWID();
    std::cout << "HWID: " << hwid << "\n\n";

    // 1) Initialize
    std::cout << "[1] Initializing...\n";
    auto init = api("init", {{"version", VERSION}});
    if (!init.value("success", false)) {
        std::cerr << "    FAILED: " << init.value("message", "Unknown error") << "\n";
        curl_global_cleanup();
        return 1;
    }
    std::cout << "    App: " << init.value("app_name", "") << " v" << init.value("version", "") << "\n";
    std::cout << "    Users: " << init.value("user_count", 0) << " | Sessions: " << init.value("active_sessions", 0) << "\n\n";

    // 2) Login
    std::string username, password;
    std::cout << "    Username: ";
    std::getline(std::cin, username);
    std::cout << "    Password: ";
    std::getline(std::cin, password);

    std::cout << "\n[2] Logging in as '" << username << "'...\n";
    auto login = api("login", {{"username", username}, {"password", password}, {"hwid", hwid}});
    if (!login.value("success", false)) {
        std::cerr << "    FAILED: " << login.value("message", "Unknown error") << "\n";
        curl_global_cleanup();
        return 1;
    }
    std::string token = login.value("token", "");
    std::cout << "    Logged in! Token: " << token.substr(0, 12) << "...\n";
    std::cout << "    Expiry: " << login.value("expiry", "N/A") << "\n\n";

    // 3) Check session
    std::cout << "[3] Validating session...\n";
    auto check = api("check", {{"token", token}});
    if (!check.value("success", false)) {
        std::cerr << "    FAILED: " << check.value("message", "Unknown error") << "\n";
        curl_global_cleanup();
        return 1;
    }
    std::cout << "    Session valid for: " << check.value("username", "") << "\n\n";

    // 4) Fetch variables
    std::cout << "[4] Fetching variables...\n";
    auto vars = api("var");
    if (vars.value("success", false) && vars.contains("variables")) {
        for (auto& [k, v] : vars["variables"].items()) {
            std::cout << "    " << k << " = " << v << "\n";
        }
    } else {
        std::cout << "    No variables found.\n";
    }

    std::cout << "\nAuthentication successful!\n";

    curl_global_cleanup();
    return 0;
}